src/object/cli_obj.c

```diff
@@ -2198,12 +2225,17 @@ obj_req_fanout(struct dc_object *obj, struct obj_auxi_args *obj_auxi,
 	struct shard_auxi_args	*shard_auxi;
 	struct daos_shard_tgt	*tgt;
 	uint32_t		 i, tgts_nr;
+	bool			 require_shard_task = false;
```

Comment:
so when we have a fanout of enumerate or key_query task, do we need to have 1 task go first and update the other tasks?

Reply:
Currently, when such an operation is invoked with DAOS_TX_NONE, their fanout RPCs still go concurrently. (These fanout RPCs may effectively see an inconsistent set of results. We would need to open an implicit TX to guarantee consistency.) When an operation is invoked with a TX handle, yes, one of the fanout RPCs goes first, and the others need to wait for this first RPC to choose an epoch. See shard_io_task and shard_query_key_task.

Reply:
i see. could you please point me to where we add the dependency on this first task that goes ahead? maybe it was done in a previous patch and i missed it.
For the TX_NONE case, I think doing an implicit TX on the client side and having to wait for one shard to go ahead first might be bad for performance in that mode. but i know this is unrelated to this PR.

Reply:
For dependency, it is in this PR. See shard_io_task and shard_query_key_task, which call dc_tx_get_epoch, who handles the dependency (tracked by dc_tx.tx_epoch_task).

For TX_NONE and fanout, it depends on the consistency we support. If we do need to support consistency, I'm afraid there's few options.

Reply:
ah i see it now. sorry about that!

Reply:
No problem at all. The PR isn't small.